cclient/iota_client_extended_api.c

```diff
@@ -0,0 +1,232 @@
+         addr_index++) {
+      tmp = iota_flex_sign_address_gen(seed, addr_index, addr_opt.security);
+      if (tmp) {
+        ret = hash243_queue_push(out_addresses, tmp);
+        if (ret) {
+          goto done;
+        }
+        free(tmp);
+      } else {
+        // gen address failed.
+        ret = RC_CCLIENT_NULL_PTR;
+        goto done;
+      }
+    }
+  } else {  // return addresses include the latest unused address.
+    for (addr_index = 0;; addr_index++) {
+      tmp = iota_flex_sign_address_gen(seed, addr_index, addr_opt.security);
+      if (tmp) {
+        ret = hash243_queue_push(out_addresses, tmp);
+        if (ret) {
+          goto done;
+        }
+        ret = is_unused_address(serv, tmp, &is_unused);
+        if (ret || is_unused) {
+          goto done;
+        }
+        free(tmp);
+      } else {
+        // gen address failed.
+        ret = RC_CCLIENT_NULL_PTR;
+        goto done;
+      }
+    }
+  }
+done:
+  if (tmp) {
+    free(tmp);
+  }
+  return ret;
+}
+
+retcode_t iota_client_get_inputs(iota_client_service_t const* const serv,
+                                 flex_trit_t const* const seed,
+                                 address_opt_t const addr_opt,
+                                 size_t const threshold,
+                                 inputs_t* const out_input) {
+  retcode_t ret_code = RC_OK;
+
+  log_debug(CCLIENT_EXTENDED_LOGGER_ID, "[%s:%d]\n", __func__, __LINE__);
+  // get address list
+  ret_code =
+      iota_client_get_new_address(serv, seed, addr_opt, &out_input->addresses);
+  if (ret_code) {
+    return ret_code;
+  }
+
+  // get balances from all addresses
+  get_balances_req_t* blances_req = get_balances_req_new();
+  get_balances_res_t* blances_res = get_balances_res_new();
+  blances_req->threshold = threshold;
+  // assign the list directly, not making a copy.
+  blances_req->addresses = out_input->addresses;
+
+  ret_code = iota_client_get_balances(serv, blances_req, blances_res);
+  out_input->total_balance = get_balances_res_total_balance(blances_res);
+
+  blances_req->addresses = NULL;  // no need to be freed
+  get_balances_req_free(&blances_req);
+  get_balances_res_free(&blances_res);
+  return ret_code;
+}
+
+retcode_t iota_client_get_account_data(iota_client_service_t const* const serv,
+                                       flex_trit_t const* const seed,
+                                       size_t const security,
+                                       account_data_t* out_account) {
+  retcode_t ret_code = RC_OK;
+  address_opt_t const addr_opt = {.security = security, .start = 0, .total = 0};
+  flex_trit_t* tmp_addr = NULL;
+  size_t addr_index = 0;
+  find_transactions_req_t* find_tx_req = NULL;
+  find_transactions_res_t* find_tx_res = NULL;
+  get_balances_req_t* blances_req = get_balances_req_new();
+  get_balances_res_t* blances_res = get_balances_res_new();
+
+  // security validation
+  if (addr_opt.security <= 0 || addr_opt.security > 3) {
```

Comment:
`addr_opt.security` is a `size_t`, it can't be `< 0`, maybe that's generating a warning